### PR TITLE
qsvdec: add support for HW_DEVICE_CTX method

### PR DESCRIPTION
**Pull request by xhaihao**
This allows user set hw_device_ctx instead of hw_frames_ctx for QSV
decoders, hence we may remove the ad-hoc libmfx setup code from FFmpeg.

"-hwaccel_output_format format" is applied to QSV decoders after
removing the ad-hoc libmfx code. To keep compatibility with old
commandlines, the default format is set to AV_PIX_FMT_QSV. User should
set "-hwaccel_output_format qsv" explicitly if AV_PIX_FMT_QSV is
expected.

User may use the standard way to set device for QSV decoders now.
"-qsv_device device" is deprecated and will be removed from FFmpeg.

For example:

$> ffmpeg -init_hw_device vaapi=va:/dev/dri/card0 -init_hw_device
qsv=hw@va -hwaccel qsv -c:v h264_qsv -i input.h264 -f null -

/dev/dri/renderD128 is actually open for h264_qsv decoder without this
patch. After applying this patch, /dev/dri/card0 is used.

### CODE REVIEW

**Comment by xhaihao** [comment]
See http://ffmpeg.org/pipermail/ffmpeg-devel/2020-September/269089.html for the discussion about the normal device selection and ad-hoc setup code.

**Comment by xhaihao** [comment]
You may apply the patch below to test qsv_h264 decoder with METHOD_HW_DEVICE_CTX, the pix_fmt must be AV_PIX_FMT_QSV

$> doc/examples/hw_decode qsv input.264 /dev/null

```
diff --git a/doc/examples/hw_decode.c b/doc/examples/hw_decode.c
index 096a229c0d..b3b517f96f 100644
--- a/doc/examples/hw_decode.c
+++ b/doc/examples/hw_decode.c
@@ -74,7 +74,7 @@ static enum AVPixelFormat get_hw_format(AVCodecContext *ctx,
     return AV_PIX_FMT_NONE;
 }
 
-static int decode_write(AVCodecContext *avctx, AVPacket *packet)
+static int decode_write(AVCodecContext *avctx, AVPacket *packet, enum AVHWDeviceType type)
 {
     AVFrame *frame = NULL, *sw_frame = NULL;
     AVFrame *tmp_frame = NULL;
@@ -105,6 +105,11 @@ static int decode_write(AVCodecContext *avctx, AVPacket *packet)
             goto fail;
         }
 
+        if (type == AV_HWDEVICE_TYPE_QSV)
+            av_assert0(frame->format == AV_PIX_FMT_QSV);
+        else if (type == AV_HWDEVICE_TYPE_VAAPI)
+            av_assert0(frame->format == AV_PIX_FMT_VAAPI);
+
         if (frame->format == hw_pix_fmt) {
             /* retrieve data from GPU to CPU */
             if ((ret = av_hwframe_transfer_data(sw_frame, frame, 0)) < 0) {
@@ -191,6 +196,9 @@ int main(int argc, char *argv[])
     }
     video_stream = ret;
 
+    if (type == AV_HWDEVICE_TYPE_QSV)
+        decoder = avcodec_find_decoder_by_name("h264_qsv");
+
     for (i = 0;; i++) {
         const AVCodecHWConfig *config = avcodec_get_hw_config(decoder, i);
         if (!config) {
@@ -231,7 +239,7 @@ int main(int argc, char *argv[])
             break;
 
         if (video_stream == packet.stream_index)
-            ret = decode_write(decoder_ctx, &packet);
+            ret = decode_write(decoder_ctx, &packet, type);
 
         av_packet_unref(&packet);
     }
@@ -239,7 +247,7 @@ int main(int argc, char *argv[])
     /* flush the decoder */
     packet.data = NULL;
     packet.size = 0;
-    ret = decode_write(decoder_ctx, &packet);
+    ret = decode_write(decoder_ctx, &packet, type);
     av_packet_unref(&packet);
 
     if (output_file)
```

**Comment by xhaihao** [comment]
@xuguangxin After applying this patch, the issue you mentioned in http://ffmpeg.org/pipermail/ffmpeg-devel/2021-February/276695.html can be fixed with normal device selection.

$> ffmpeg -init_hw_device vaapi=intel:/dev/dri/renderD128 -init_hw_device qsv=qsv@intel -init_hw_device opencl=ocl@intel -filter_hw_device ocl -hwaccel qsv -c:v h264_qsv -hwaccel_output_format qsv -i input.h264 -vf 'hwmap=derive_device=opencl,format=opencl,unsharp_opencl,hwmap=derive_device=qsv:reverse=1:extra_hw_frames=32'  -c:v hevc_qsv  -y test.h265

**Comment by xuguangxin** [comment]
@xhaihao , could you try the command line suggested by @fhvwy 
 thanks 
https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210222084504.339978-1-guangxin.xu@intel.com/

**Comment by xhaihao** [comment]
The command below also works after using '-init_hw_device qsv=qsv@intel' to create the connection between vaapi device and qsv device.

$> ffmpeg -init_hw_device vaapi=intel:/dev/dri/renderD128 -init_hw_device qsv=qsv@intel -init_hw_device opencl=ocl@intel -hwaccel qsv -c:v h264_qsv -hwaccel_output_format qsv -i input.h264 -filter_hw_device ocl -vf 'hwmap,format=opencl,unsharp_opencl,hwmap=derive_device=qsv:reverse=1:extra_hw_frames=32'  -c:v hevc_qsv  -y test.h265

File: fftools/ffmpeg_opt.c [hunk 1]
```diff
@@ -569,6 +566,49 @@ static int opt_vaapi_device(void *optctx, const char *opt, const char *arg)
 }
 #endif
 
+#if CONFIG_QSV
+static int opt_qsv_device(void *optctx, const char *opt, const char *arg)
+{
+#if CONFIG_VAAPI
+    const char *prefix = "vaapi=__qsv_child_device:";
+#elif CONFIG_DXVA2
+    const char *prefix = "dxva2=__qsv_child_device:";
```
Comment (xuguangxin) [inline fftools/ffmpeg_opt.c:575]: why it has __?

Comment (xhaihao) [reply fftools/ffmpeg_opt.c:575]: __qsv_child_device is the name for a hw device. Using __ as prefix to reduce the possibility of using the same device name with user's device. 

I also added a warning of not use __qsv_child_device or __qsv_device as device name in user's commandline.

**Comment by xhaihao** [comment]
Updated the comment 

Is there any other feedback for this patch?

**Review by xuguangxin — DISMISSED** [review]

**Comment by wenbinc-Bin** [comment]
@xhaihao, Sorry for the late response. You may forget to remove the definition of HWACCEL_QSV in ffmpeg.h.

**Comment by xhaihao** [comment]
> @xhaihao, Sorry for the late response. You may forget to remove the definition of HWACCEL_QSV in ffmpeg.h.

Thanks for the comment. I am aware of this enum, but I was not sure whether this enum must be removed when submitting this PR.  Maybe it is better to remove it, I will update the patch.

**Comment by xhaihao** [comment]
Updated the PR to remove HWACCEL_QSV and added a commit to simplify doc/examples/qsvdec by using hw_device_ctx interface

**Comment by xhaihao** [comment]
sent the patches to FFmpeg mailing list and closed this PR. 

http://ffmpeg.org/pipermail/ffmpeg-devel/2021-June/281774.html
http://ffmpeg.org/pipermail/ffmpeg-devel/2021-June/281773.html

**Pull request closed without merging**